{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_analysis_folder(operating_system, project_name, analysis_name):\n",
    "\t\n",
    "\t# Set path to BSC folder for mac, linux and windows\n",
    "\tif operating_system == 'windows':\n",
    "\t\tpathToBSCfolder = 'Z:\\\\'\n",
    "\telif operating_system == 'linux':\n",
    "\t\tpathToBSCfolder = os.path.join('/home','amontaner','BSC')\n",
    "\telif operating_system == 'mac':\n",
    "\t\tpathToBSCfolder = os.path.join('/Users','annamontaner','Documents','BSC3')\n",
    "\telse:\n",
    "\t\tprint ('undefined operating system') \n",
    "\t\n",
    "\t# Path to project folder \n",
    "\tproject_folder = os.path.join('scratch','cli79','cli79334','projects', 'other', project_name)\n",
    "\n",
    "\n",
    "   # Create analysis folder and subfolders \n",
    "\tanalysis_folder = os.path.join(pathToBSCfolder, project_folder, analysis_name)\n",
    "\tinput_data = os.path.join(analysis_folder,'input_data')\n",
    "\toutput_data = os.path.join(analysis_folder,'output_data')\n",
    "\n",
    "\n",
    "\tfolder_list = [analysis_folder,\n",
    "\t\t\t\t   input_data, \n",
    "\t\t\t\t   output_data]\n",
    "\n",
    "\tif not os.path.exists(analysis_folder):\n",
    "\t\tfor items in folder_list:\n",
    "\t\t\tos.makedirs(items)\n",
    "\t\tprint(\"Analysis folder and subfolders created '{}' created\".format(analysis_name))\n",
    "\telse: \n",
    "\t\tprint(\"Already existing analysis '{}' \".format(analysis_name))\n",
    "\n",
    "\treturn(analysis_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Already existing analysis '01_vcf_generation' \n"
     ]
    }
   ],
   "source": [
    "# Create analysis folders\n",
    "analysis_folder = create_analysis_folder('mac','1kGPhg38_african_variants_abundant_specific', '01_vcf_generation')\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/kt/qf89bw917mg3nhrtrkj4_lmc0000gn/T/ipykernel_1438/1752083001.py:3: DtypeWarning: Columns (1,130) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  imported = pd.read_csv(os.path.join(path,\"African_specific.tsv\"), sep=\"\\t\")\n"
     ]
    }
   ],
   "source": [
    "#Â Import African abundant variants : it should be imported from the input folder, but now it's found in the desktop folder\n",
    "path = \"/Users/annamontaner/Desktop/African_variants/\"\n",
    "imported = pd.read_csv(os.path.join(path,\"African_specific.tsv\"), sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "106323"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check rows and columns of the imported data\n",
    "imported.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "100455"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(imported['ID']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ANN[*].GENE</th>\n",
       "      <th>ANN[*].GENEID</th>\n",
       "      <th>ANN[*].FEATURE</th>\n",
       "      <th>ANN[*].FEATUREID</th>\n",
       "      <th>ANN[*].BIOTYPE</th>\n",
       "      <th>...</th>\n",
       "      <th>dbNSFP_clinvar_Orphanet_id</th>\n",
       "      <th>dbNSFP_Interpro_domain</th>\n",
       "      <th>fedorova_label</th>\n",
       "      <th>ADME_label</th>\n",
       "      <th>abundantAFR_label</th>\n",
       "      <th>maxFreqs</th>\n",
       "      <th>AFR_overrepresentation</th>\n",
       "      <th>AFR_overrepresentation_mod</th>\n",
       "      <th>AFR_overrepresentation_label</th>\n",
       "      <th>specificAFR_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>67181</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Not_AFR_ARSA_Fedorova</td>\n",
       "      <td>Not_ADME</td>\n",
       "      <td>abundant</td>\n",
       "      <td>0.02</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>10</td>\n",
       "      <td>specific</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>927303</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000616016.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Not_AFR_ARSA_Fedorova</td>\n",
       "      <td>Not_ADME</td>\n",
       "      <td>abundant</td>\n",
       "      <td>0.02</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>10</td>\n",
       "      <td>specific</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>931526</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000616016.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Not_AFR_ARSA_Fedorova</td>\n",
       "      <td>Not_ADME</td>\n",
       "      <td>abundant</td>\n",
       "      <td>0.02</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>10</td>\n",
       "      <td>specific</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>931990</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000616016.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Not_AFR_ARSA_Fedorova</td>\n",
       "      <td>Not_ADME</td>\n",
       "      <td>abundant</td>\n",
       "      <td>0.02</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>10</td>\n",
       "      <td>specific</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>932038</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000616016.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Not_AFR_ARSA_Fedorova</td>\n",
       "      <td>Not_ADME</td>\n",
       "      <td>abundant</td>\n",
       "      <td>0.01</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>20</td>\n",
       "      <td>specific</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106318</th>\n",
       "      <td>chrX</td>\n",
       "      <td>X</td>\n",
       "      <td>2767152</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>XG</td>\n",
       "      <td>ENSG00000124343</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000644266.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Not_AFR_ARSA_Fedorova</td>\n",
       "      <td>Not_ADME</td>\n",
       "      <td>abundant</td>\n",
       "      <td>0.03</td>\n",
       "      <td>8.333333</td>\n",
       "      <td>8.333333</td>\n",
       "      <td>8</td>\n",
       "      <td>specific</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106319</th>\n",
       "      <td>chrX</td>\n",
       "      <td>X</td>\n",
       "      <td>2775070</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>XG</td>\n",
       "      <td>ENSG00000124343</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000644266.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Not_AFR_ARSA_Fedorova</td>\n",
       "      <td>Not_ADME</td>\n",
       "      <td>abundant</td>\n",
       "      <td>0.02</td>\n",
       "      <td>11.500000</td>\n",
       "      <td>11.500000</td>\n",
       "      <td>12</td>\n",
       "      <td>specific</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106320</th>\n",
       "      <td>chrX</td>\n",
       "      <td>X</td>\n",
       "      <td>2775075</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>XG</td>\n",
       "      <td>ENSG00000124343</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000644266.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Not_AFR_ARSA_Fedorova</td>\n",
       "      <td>Not_ADME</td>\n",
       "      <td>abundant</td>\n",
       "      <td>0.02</td>\n",
       "      <td>11.500000</td>\n",
       "      <td>11.500000</td>\n",
       "      <td>12</td>\n",
       "      <td>specific</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106321</th>\n",
       "      <td>chrX</td>\n",
       "      <td>X</td>\n",
       "      <td>156001085</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>ENSG00000124334</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000244174.11</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Not_AFR_ARSA_Fedorova</td>\n",
       "      <td>Not_ADME</td>\n",
       "      <td>abundant</td>\n",
       "      <td>0.02</td>\n",
       "      <td>11.500000</td>\n",
       "      <td>11.500000</td>\n",
       "      <td>12</td>\n",
       "      <td>specific</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106322</th>\n",
       "      <td>chrX</td>\n",
       "      <td>X</td>\n",
       "      <td>156002438</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>ENSG00000124334</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000244174.11</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>Not_AFR_ARSA_Fedorova</td>\n",
       "      <td>Not_ADME</td>\n",
       "      <td>abundant</td>\n",
       "      <td>0.02</td>\n",
       "      <td>10.500000</td>\n",
       "      <td>10.500000</td>\n",
       "      <td>10</td>\n",
       "      <td>specific</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106323 rows Ã 152 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       SAMPLE CHROM        POS REF ALT ANN[*].GENE    ANN[*].GENEID  \\\n",
       "0        chr1     1      67181   A   G       OR4F5  ENSG00000186092   \n",
       "1        chr1     1     927303   G   A      SAMD11  ENSG00000187634   \n",
       "2        chr1     1     931526   G   A      SAMD11  ENSG00000187634   \n",
       "3        chr1     1     931990   T   C      SAMD11  ENSG00000187634   \n",
       "4        chr1     1     932038   C   T      SAMD11  ENSG00000187634   \n",
       "...       ...   ...        ...  ..  ..         ...              ...   \n",
       "106318   chrX     X    2767152   G   A          XG  ENSG00000124343   \n",
       "106319   chrX     X    2775070   T   C          XG  ENSG00000124343   \n",
       "106320   chrX     X    2775075   C   A          XG  ENSG00000124343   \n",
       "106321   chrX     X  156001085   T   C        IL9R  ENSG00000124334   \n",
       "106322   chrX     X  156002438   A   G        IL9R  ENSG00000124334   \n",
       "\n",
       "       ANN[*].FEATURE    ANN[*].FEATUREID  ANN[*].BIOTYPE  ...  \\\n",
       "0          transcript   ENST00000641515.2  protein_coding  ...   \n",
       "1          transcript   ENST00000616016.5  protein_coding  ...   \n",
       "2          transcript   ENST00000616016.5  protein_coding  ...   \n",
       "3          transcript   ENST00000616016.5  protein_coding  ...   \n",
       "4          transcript   ENST00000616016.5  protein_coding  ...   \n",
       "...               ...                 ...             ...  ...   \n",
       "106318     transcript   ENST00000644266.2  protein_coding  ...   \n",
       "106319     transcript   ENST00000644266.2  protein_coding  ...   \n",
       "106320     transcript   ENST00000644266.2  protein_coding  ...   \n",
       "106321     transcript  ENST00000244174.11  protein_coding  ...   \n",
       "106322     transcript  ENST00000244174.11  protein_coding  ...   \n",
       "\n",
       "       dbNSFP_clinvar_Orphanet_id dbNSFP_Interpro_domain  \\\n",
       "0                               .                      .   \n",
       "1                               .                      .   \n",
       "2                               .                      .   \n",
       "3                               .                      .   \n",
       "4                               .                      .   \n",
       "...                           ...                    ...   \n",
       "106318                          .                      .   \n",
       "106319                          .                      .   \n",
       "106320                          .                      .   \n",
       "106321                          .                      .   \n",
       "106322                          .                      .   \n",
       "\n",
       "               fedorova_label ADME_label abundantAFR_label  maxFreqs  \\\n",
       "0       Not_AFR_ARSA_Fedorova   Not_ADME          abundant      0.02   \n",
       "1       Not_AFR_ARSA_Fedorova   Not_ADME          abundant      0.02   \n",
       "2       Not_AFR_ARSA_Fedorova   Not_ADME          abundant      0.02   \n",
       "3       Not_AFR_ARSA_Fedorova   Not_ADME          abundant      0.02   \n",
       "4       Not_AFR_ARSA_Fedorova   Not_ADME          abundant      0.01   \n",
       "...                       ...        ...               ...       ...   \n",
       "106318  Not_AFR_ARSA_Fedorova   Not_ADME          abundant      0.03   \n",
       "106319  Not_AFR_ARSA_Fedorova   Not_ADME          abundant      0.02   \n",
       "106320  Not_AFR_ARSA_Fedorova   Not_ADME          abundant      0.02   \n",
       "106321  Not_AFR_ARSA_Fedorova   Not_ADME          abundant      0.02   \n",
       "106322  Not_AFR_ARSA_Fedorova   Not_ADME          abundant      0.02   \n",
       "\n",
       "        AFR_overrepresentation  AFR_overrepresentation_mod  \\\n",
       "0                    10.000000                   10.000000   \n",
       "1                    10.000000                   10.000000   \n",
       "2                    10.000000                   10.000000   \n",
       "3                    10.000000                   10.000000   \n",
       "4                    20.000000                   20.000000   \n",
       "...                        ...                         ...   \n",
       "106318                8.333333                    8.333333   \n",
       "106319               11.500000                   11.500000   \n",
       "106320               11.500000                   11.500000   \n",
       "106321               11.500000                   11.500000   \n",
       "106322               10.500000                   10.500000   \n",
       "\n",
       "        AFR_overrepresentation_label  specificAFR_label  \n",
       "0                                 10           specific  \n",
       "1                                 10           specific  \n",
       "2                                 10           specific  \n",
       "3                                 10           specific  \n",
       "4                                 20           specific  \n",
       "...                              ...                ...  \n",
       "106318                             8           specific  \n",
       "106319                            12           specific  \n",
       "106320                            12           specific  \n",
       "106321                            12           specific  \n",
       "106322                            10           specific  \n",
       "\n",
       "[106323 rows x 152 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imported"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copy imported dataframe\n",
    "afr_specific = imported.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build columns required for vcf\n",
    "afr_specific['QUAL'] = \".\"\n",
    "afr_specific['FILTER'] = \"NONE\"\n",
    "afr_specific['#CHROM'] = afr_specific['CHROM']\n",
    "afr_specific['INFO'] = 'fedorova_label='+afr_specific['fedorova_label']+\";adme_label=\"+afr_specific['ADME_label']+\";abundantAFR_label=\"+afr_specific['abundantAFR_label']+\";maxFreqs=\"+afr_specific['maxFreqs'].astype(str)+\";AFR_overrepresentation=\"+afr_specific['AFR_overrepresentation'].astype(str)+\";AFR_overrepresentation_mod=\"+afr_specific['AFR_overrepresentation_mod'].astype(str)+\";AFR_overrepresentation_label=\"+afr_specific['AFR_overrepresentation_label'].astype(str)+\";specificAFR_label=\"+afr_specific['specificAFR_label']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ANN[*].GENE</th>\n",
       "      <th>ANN[*].GENEID</th>\n",
       "      <th>ANN[*].FEATURE</th>\n",
       "      <th>ANN[*].FEATUREID</th>\n",
       "      <th>ANN[*].BIOTYPE</th>\n",
       "      <th>...</th>\n",
       "      <th>abundantAFR_label</th>\n",
       "      <th>maxFreqs</th>\n",
       "      <th>AFR_overrepresentation</th>\n",
       "      <th>AFR_overrepresentation_mod</th>\n",
       "      <th>AFR_overrepresentation_label</th>\n",
       "      <th>specificAFR_label</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>INFO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>67181</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>abundant</td>\n",
       "      <td>0.02</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>10</td>\n",
       "      <td>specific</td>\n",
       "      <td>.</td>\n",
       "      <td>NONE</td>\n",
       "      <td>1</td>\n",
       "      <td>fedorova_label=Not_AFR_ARSA_Fedorova;adme_labe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>927303</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000616016.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>abundant</td>\n",
       "      <td>0.02</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>10</td>\n",
       "      <td>specific</td>\n",
       "      <td>.</td>\n",
       "      <td>NONE</td>\n",
       "      <td>1</td>\n",
       "      <td>fedorova_label=Not_AFR_ARSA_Fedorova;adme_labe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>931526</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000616016.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>abundant</td>\n",
       "      <td>0.02</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>10</td>\n",
       "      <td>specific</td>\n",
       "      <td>.</td>\n",
       "      <td>NONE</td>\n",
       "      <td>1</td>\n",
       "      <td>fedorova_label=Not_AFR_ARSA_Fedorova;adme_labe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>931990</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000616016.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>abundant</td>\n",
       "      <td>0.02</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>10</td>\n",
       "      <td>specific</td>\n",
       "      <td>.</td>\n",
       "      <td>NONE</td>\n",
       "      <td>1</td>\n",
       "      <td>fedorova_label=Not_AFR_ARSA_Fedorova;adme_labe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>932038</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000616016.5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>abundant</td>\n",
       "      <td>0.01</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>20</td>\n",
       "      <td>specific</td>\n",
       "      <td>.</td>\n",
       "      <td>NONE</td>\n",
       "      <td>1</td>\n",
       "      <td>fedorova_label=Not_AFR_ARSA_Fedorova;adme_labe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106318</th>\n",
       "      <td>chrX</td>\n",
       "      <td>X</td>\n",
       "      <td>2767152</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>XG</td>\n",
       "      <td>ENSG00000124343</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000644266.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>abundant</td>\n",
       "      <td>0.03</td>\n",
       "      <td>8.333333</td>\n",
       "      <td>8.333333</td>\n",
       "      <td>8</td>\n",
       "      <td>specific</td>\n",
       "      <td>.</td>\n",
       "      <td>NONE</td>\n",
       "      <td>X</td>\n",
       "      <td>fedorova_label=Not_AFR_ARSA_Fedorova;adme_labe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106319</th>\n",
       "      <td>chrX</td>\n",
       "      <td>X</td>\n",
       "      <td>2775070</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>XG</td>\n",
       "      <td>ENSG00000124343</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000644266.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>abundant</td>\n",
       "      <td>0.02</td>\n",
       "      <td>11.500000</td>\n",
       "      <td>11.500000</td>\n",
       "      <td>12</td>\n",
       "      <td>specific</td>\n",
       "      <td>.</td>\n",
       "      <td>NONE</td>\n",
       "      <td>X</td>\n",
       "      <td>fedorova_label=Not_AFR_ARSA_Fedorova;adme_labe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106320</th>\n",
       "      <td>chrX</td>\n",
       "      <td>X</td>\n",
       "      <td>2775075</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>XG</td>\n",
       "      <td>ENSG00000124343</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000644266.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>abundant</td>\n",
       "      <td>0.02</td>\n",
       "      <td>11.500000</td>\n",
       "      <td>11.500000</td>\n",
       "      <td>12</td>\n",
       "      <td>specific</td>\n",
       "      <td>.</td>\n",
       "      <td>NONE</td>\n",
       "      <td>X</td>\n",
       "      <td>fedorova_label=Not_AFR_ARSA_Fedorova;adme_labe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106321</th>\n",
       "      <td>chrX</td>\n",
       "      <td>X</td>\n",
       "      <td>156001085</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>ENSG00000124334</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000244174.11</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>abundant</td>\n",
       "      <td>0.02</td>\n",
       "      <td>11.500000</td>\n",
       "      <td>11.500000</td>\n",
       "      <td>12</td>\n",
       "      <td>specific</td>\n",
       "      <td>.</td>\n",
       "      <td>NONE</td>\n",
       "      <td>X</td>\n",
       "      <td>fedorova_label=Not_AFR_ARSA_Fedorova;adme_labe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106322</th>\n",
       "      <td>chrX</td>\n",
       "      <td>X</td>\n",
       "      <td>156002438</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>ENSG00000124334</td>\n",
       "      <td>transcript</td>\n",
       "      <td>ENST00000244174.11</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>abundant</td>\n",
       "      <td>0.02</td>\n",
       "      <td>10.500000</td>\n",
       "      <td>10.500000</td>\n",
       "      <td>10</td>\n",
       "      <td>specific</td>\n",
       "      <td>.</td>\n",
       "      <td>NONE</td>\n",
       "      <td>X</td>\n",
       "      <td>fedorova_label=Not_AFR_ARSA_Fedorova;adme_labe...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106323 rows Ã 156 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       SAMPLE CHROM        POS REF ALT ANN[*].GENE    ANN[*].GENEID  \\\n",
       "0        chr1     1      67181   A   G       OR4F5  ENSG00000186092   \n",
       "1        chr1     1     927303   G   A      SAMD11  ENSG00000187634   \n",
       "2        chr1     1     931526   G   A      SAMD11  ENSG00000187634   \n",
       "3        chr1     1     931990   T   C      SAMD11  ENSG00000187634   \n",
       "4        chr1     1     932038   C   T      SAMD11  ENSG00000187634   \n",
       "...       ...   ...        ...  ..  ..         ...              ...   \n",
       "106318   chrX     X    2767152   G   A          XG  ENSG00000124343   \n",
       "106319   chrX     X    2775070   T   C          XG  ENSG00000124343   \n",
       "106320   chrX     X    2775075   C   A          XG  ENSG00000124343   \n",
       "106321   chrX     X  156001085   T   C        IL9R  ENSG00000124334   \n",
       "106322   chrX     X  156002438   A   G        IL9R  ENSG00000124334   \n",
       "\n",
       "       ANN[*].FEATURE    ANN[*].FEATUREID  ANN[*].BIOTYPE  ...  \\\n",
       "0          transcript   ENST00000641515.2  protein_coding  ...   \n",
       "1          transcript   ENST00000616016.5  protein_coding  ...   \n",
       "2          transcript   ENST00000616016.5  protein_coding  ...   \n",
       "3          transcript   ENST00000616016.5  protein_coding  ...   \n",
       "4          transcript   ENST00000616016.5  protein_coding  ...   \n",
       "...               ...                 ...             ...  ...   \n",
       "106318     transcript   ENST00000644266.2  protein_coding  ...   \n",
       "106319     transcript   ENST00000644266.2  protein_coding  ...   \n",
       "106320     transcript   ENST00000644266.2  protein_coding  ...   \n",
       "106321     transcript  ENST00000244174.11  protein_coding  ...   \n",
       "106322     transcript  ENST00000244174.11  protein_coding  ...   \n",
       "\n",
       "       abundantAFR_label maxFreqs  AFR_overrepresentation  \\\n",
       "0               abundant     0.02               10.000000   \n",
       "1               abundant     0.02               10.000000   \n",
       "2               abundant     0.02               10.000000   \n",
       "3               abundant     0.02               10.000000   \n",
       "4               abundant     0.01               20.000000   \n",
       "...                  ...      ...                     ...   \n",
       "106318          abundant     0.03                8.333333   \n",
       "106319          abundant     0.02               11.500000   \n",
       "106320          abundant     0.02               11.500000   \n",
       "106321          abundant     0.02               11.500000   \n",
       "106322          abundant     0.02               10.500000   \n",
       "\n",
       "       AFR_overrepresentation_mod AFR_overrepresentation_label  \\\n",
       "0                       10.000000                           10   \n",
       "1                       10.000000                           10   \n",
       "2                       10.000000                           10   \n",
       "3                       10.000000                           10   \n",
       "4                       20.000000                           20   \n",
       "...                           ...                          ...   \n",
       "106318                   8.333333                            8   \n",
       "106319                  11.500000                           12   \n",
       "106320                  11.500000                           12   \n",
       "106321                  11.500000                           12   \n",
       "106322                  10.500000                           10   \n",
       "\n",
       "        specificAFR_label  QUAL  FILTER  #CHROM  \\\n",
       "0                specific     .    NONE       1   \n",
       "1                specific     .    NONE       1   \n",
       "2                specific     .    NONE       1   \n",
       "3                specific     .    NONE       1   \n",
       "4                specific     .    NONE       1   \n",
       "...                   ...   ...     ...     ...   \n",
       "106318           specific     .    NONE       X   \n",
       "106319           specific     .    NONE       X   \n",
       "106320           specific     .    NONE       X   \n",
       "106321           specific     .    NONE       X   \n",
       "106322           specific     .    NONE       X   \n",
       "\n",
       "                                                     INFO  \n",
       "0       fedorova_label=Not_AFR_ARSA_Fedorova;adme_labe...  \n",
       "1       fedorova_label=Not_AFR_ARSA_Fedorova;adme_labe...  \n",
       "2       fedorova_label=Not_AFR_ARSA_Fedorova;adme_labe...  \n",
       "3       fedorova_label=Not_AFR_ARSA_Fedorova;adme_labe...  \n",
       "4       fedorova_label=Not_AFR_ARSA_Fedorova;adme_labe...  \n",
       "...                                                   ...  \n",
       "106318  fedorova_label=Not_AFR_ARSA_Fedorova;adme_labe...  \n",
       "106319  fedorova_label=Not_AFR_ARSA_Fedorova;adme_labe...  \n",
       "106320  fedorova_label=Not_AFR_ARSA_Fedorova;adme_labe...  \n",
       "106321  fedorova_label=Not_AFR_ARSA_Fedorova;adme_labe...  \n",
       "106322  fedorova_label=Not_AFR_ARSA_Fedorova;adme_labe...  \n",
       "\n",
       "[106323 rows x 156 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "afr_specific"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate .tsv to export\n",
    "afr_specific_for_vcf = afr_specific[['#CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export .tsv\n",
    "afr_specific_for_vcf.to_csv(os.path.join(analysis_folder,'output_data','afr_specific_for_vcf.tsv'), index=False, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Generate vcf for SnpEff annotation\n",
    "input_TSV = os.path.join(os.path.join(analysis_folder,'output_data',\"afr_specific_for_vcf.tsv\"))\n",
    "output_VCF = os.path.join(os.path.join(analysis_folder,'output_data',\"afr_specific.vcf\"))\n",
    "\n",
    "I = open(input_TSV, \"r\")\n",
    "O = open(output_VCF, \"w\")\n",
    "\n",
    "O.write(\"##fileformat=VCFv4.0\\n\")\n",
    "O.write(\"##FILTER=<ID=ID,Number=1,Type=String,Description=\\\"dbSnp id\\\">\\n\")\n",
    "O.write(\"##INFO=<ID=fedorova_label,Number=1,Type=String,Description=\\\"Present or not present in Fedorova data\\\">\\n\")\n",
    "O.write(\"##INFO=<ID=adme_label,Number=1,Type=String,Description=\\\"ADME or not ADME genes\\\">\\n\")\n",
    "O.write(\"##INFO=<ID=abundantAFR_label,Number=1,Type=String,Description=\\\"Variant status: abundant or not abundant in African populations in the 1kGPhg38 dataset\\\">\\n\")\n",
    "O.write(\"##INFO=<ID=maxFreqs,Number=1,Type=Float,Description=\\\"Maximum frequency observed in populations other than African in the 1kGPhg38 dataset\\\">\\n\")\n",
    "O.write(\"##INFO=<ID=AFR_overrepresentation,Number=A,Type=Float,Description=\\\"African specificity raw\\\">\\n\")\n",
    "O.write(\"##INFO=<ID=AFR_overrepresentation_mod,Number=A,Type=Float,Description=\\\"African specificity modified\\\">\\n\")\n",
    "O.write(\"##INFO=<ID=AFR_overrepresentation_label,Number=1,Type=String,Description=\\\"African specificity as factor\\\">\\n\")\n",
    "O.write(\"##INFO=<ID=specificAFR_label,Number=1,Type=String,Description=\\\"Variant status: specific or not specific to African populations in the 1kGPhg38 dataset\\\">\\n\")\n",
    "O.write(\"##FORMAT=<ID=NO,Number=1,Type=String,Description=\\\"Not used\\\">\\n\")\n",
    "\n",
    "\n",
    "for iLine in I:\n",
    "\tlist = iLine.split(\"\\t\")\n",
    "\tnewLine = '\\t'.join(list)\n",
    "\tO.write(newLine)\n",
    "\n",
    "I.close()\n",
    "O.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ersilia",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
